pyanaconda/modules/payloads/payload/dnf/dnf_manager.py

```diff
@@ -616,17 +709,46 @@ def download_packages(self, callback):
         :param callback: a callback for progress reporting
         :raise PayloadInstallationError: if the download fails
         """
-        packages = self._base.transaction.install_set  # pylint: disable=no-member
-        progress = DownloadProgress(callback=callback)
+        # Set up the download callbacks.
+        progress = DownloadProgress(callback)
+        self._set_download_callbacks(progress)
+
+        # Prepare packages for download.
+        downloader = libdnf5.repo.PackageDownloader(self._base)
+        packages = self._get_download_packages()
+        destination = self.download_location
 
-        log.info("Downloading packages to %s.", self.download_location)
+        for package in packages:
+            downloader.add(package, destination)
+
+        downloader.set_fail_fast(True)
+        downloader.set_resume(False)
+
+        # Download the packages.
+        log.info("Downloading packages to %s.", destination)
 
         try:
-            self._base.download_packages(packages, progress)
-        except dnf.exceptions.DownloadError as e:
+            downloader.download()
+        except (libdnf5.exception.Error, libdnf5.exception.NonLibdnf5Exception) as e:
             msg = "Failed to download the following packages: " + str(e)
             raise PayloadInstallationError(msg) from None
 
+    def _set_download_callbacks(self, callbacks):
+        """Set up the download callbacks."""
+        self._base.set_download_callbacks(
+            libdnf5.repo.DownloadCallbacksUniquePtr(callbacks)
+        )
+
+    def _get_download_packages(self):
+        """Get a list of resolved packages to download."""
+        if not self._transaction:
+            raise RuntimeError("There is no transaction to use!")
+
+        return [
+            tspkg.get_package() for tspkg in self._transaction.get_transaction_packages()
+            if libdnf5.base.transaction.transaction_item_action_is_inbound(tspkg.get_action())
```

Comment:
What is 'if libdnf5.base.transaction.transaction_item_action_is_inbound(tspkg.get_action())' doing?

Reply:
I added a comment that should explain that. I added it in an extra commit, because this function was completely written by Vendy and I don't want to alter her initial commit.